context.go

```diff
@@ -321,42 +306,40 @@ func (c *context) Apply(resource Resource) error {
 		}
 
 	case Directory:
-		if !exists {
+		if fi == nil {
 			if err := c.driver.Mkdir(fp, resource.Mode()); err != nil {
 				return err
 			}
 		}
 	case SymLink:
-		target, err := c.resolveSymlink(r)
-		if err != nil {
-			return err
-		}
-		if exists {
-			currentPath, err := c.driver.Readlink(fp)
-			if err != nil {
-				return err
-			}
-			if currentPath != target {
-				if err := c.driver.Remove(fp); err != nil {
+		var target string // only possibly set if target resource is a symlink
+
+		if fi != nil {
+			if fi.Mode()&os.ModeSymlink != 0 {
+				target, err = c.driver.Readlink(fp)
+				if err != nil {
 					return err
 				}
-				exists = false
 			}
 		}
-		if !exists {
-			if err := c.driver.Symlink(target, fp); err != nil {
+
+		if target != r.Target() {
+			if err := c.driver.Remove(fp); err != nil { // RemoveAll?
+				return err
+			}
+
+			if err := c.driver.Symlink(r.Target(), fp); err != nil {
 				return err
 			}
-			// Not supported on linux, skip chmod on links
-			//if err := c.driver.Lchmod(fp, resource.Mode()); err != nil {
-			//	return err
-			//}
 		}
+
+		// NOTE(stevvooe): Chmod on symlink is not supported on linux. We
+		// may want to maintain support for other platforms that have it.
 		chmod = false
 	}
 
 	// Update filemode if file was not created
-	if chmod && exists {
+	if chmod {
```

Comment:
This code needs some general syscall optimization but I wanted to avoid the refactoring to get there.